### ISSUE TITLE
print: problem with printf formatters for _Decimal datatypes, 

### ISSUE THREAD

**Issue by newbie-02** [libdfp/libdfp#186]
hello @all, pls. be tolerant, I'm not a pro and even can't say if it's better reporting  
here or at gcc.  
  
I experienced difficulties in printing _Decimal datatypes, e.g. '%.25Df' doesn't  
work / isn't evaluated by gcc unless I  
1.) link against libdfp, AND! 
2.) have a assignment  of a _Decimalxxx value to a variable - commented III. below,  
and use it in a printf statement - commented IV. below.  
  
I'd put it in a little test program, see comments there, un-uncommenting the  
rows marked III. and IV. and compiling with -ldfp brings the wanted output.  
  
Reported as I can imagine that not being able to print is a blocker for others  
using _Decimal, and finding this workaround is too tricky that plenty will do.  
  
TIA for any help / explanaion ...     
```
// simple test program to check if printing '_Decimalxxx' datatypes works, 
 
#include <stdio.h> 									// reg. e.g. printf, 

// it's a weird behavior how using the libdfp library works, here compiling with 
// -ldfp is not sufficient to make I. andd II. below work, 
// but having III. and IV. active helps to get expected print formatting for I. and II. 
// don't ask me why!, it is puzzled together by trial and error ... 
 
// compiled with 'gcc -o test1 test1.c -ldfp' and run with './test1' I get: 
// _Decimal64 0.2 printed as decimal:    %.25Df 
// _Decimal64 0.2 printed as scientific: %.25DE 
 
// un-uncommenting the rows marked III. and IV and compiled with 'gcc -o test1 test1.c -ldfp' I get: 
// _Decimal64 0.2 printed as decimal:    0.2000000000000000000000000 
// _Decimal64 0.2 printed as scientific: 2.0000000000000000000000000E-01 
// _Decimal64 0.2 printed as double:     0.2000000000000000111022302 
 
// rows marked III. and IV active and compiled without linking to libdfp 'gcc -o test1 test1.c' I get: 
// _Decimal64 0.2 printed as decimal:    %.25Df 
// _Decimal64 0.2 printed as scientific: %.25DE 
// _Decimal64 0.2 printed as double:     0.2000000000000000111022302 
 
// I don't know what the following line is for, just tested if it helps for the printing problem, seems not, 
#define __STDC_WANT_DEC_FP__ 1 								// testing if it helps in print formatting, 
 
int main() { 
	printf("_Decimal64 0.2 printed as decimal:    %.25Df\n", 0.2DD ); 		// I. 
	printf("_Decimal64 0.2 printed as scientific: %.25DE\n", 0.2DD ); 		// I!. 
 
// 	_Decimal64 num1 = 0.2DD; 							// III. 
// 	printf("_Decimal64 0.2 printed as double:     %.25f\n", (double)num1 ); 	// IV. 
 
	return 0; 
} 

```

**Comment by pmur** [comment]
Hi, there are two decimal float features you are using.

1. _Decimal64 (and others) provided by gcc.
2. The printf feature to print _Decimal values.

The second feature is passively enabled by linking libdfp. It works via a glibc mechanism called [printf hooks](https://www.gnu.org/software/libc/manual/html_node/Customizing-Printf.html). If libdfp is not linked with the binary, %Df and other decimal related modifiers will not work. libdfp's [README.user](https://github.com/libdfp/libdfp/blob/master/README.user#L262) describes the supported printf hooks.

In short, to use the printf support for _Decimal types, the binary must be linked against libdfp. There is no compile time enforcement of this requirement.

**Comment by newbie-02** [comment]
hello @pmur and @ryanarn, thanks for your comments,  
  
> The second feature is passively enabled by linking libdfp  
  
understood and experienced that, the funny thing is that it does not! work  
with the test program as quoted in my OP, but needs having the rows marked  
III. and IV. active.  
  
> README.user  
  
massively contributed to my confusion, but didn't explain the question above.  
'gcc -v 2>&1 | grep "\-\-enable\-decimal\-float" ' resultet in an empty answer  
here.  
  
> In short, to use the printf support for _Decimal types, the binary must be linked against libdfp.  
  
think I did so by compiling with -ldfp, or is there any other requirement?  
> should define the __STDC_WANT_DEC_FP__  
  
as you can see in my test / sample program I also tried that, with no change,  
compiling with -ldfp still isn't sufficient to get  'decimal print formatters' to  
work, it requires the two lines in the code ( others may help also ).  
  
Would be nice if you check the above program on your system, and if it doesn't  
work try what it's dependent on, if it works try to estimate what may be  
misconfigured on my system.  
  
TIA for any help.

**Comment by pmur** [comment]
I will need more information about your development environment to help. Some helpful bits would be the linux distro you are using, how you are invoking gcc on your example, the version of gcc, the version of libdfp, how libdfp was built.

For your example, no macros need to be defined to use the _Decimal types, or printf.

For example on my Fedora 38 amd64 machine (with libdfp-devel installed via dnf), with your example saved too foo.c:
```
gcc foo.c -ldfp && ./a.out 
_Decimal64 0.2 printed as decimal:    0.2000000000000000000000000
_Decimal64 0.2 printed as scientific: 2.0000000000000000000000000E-01
```
and
```
gcc foo.c && ./a.out 
_Decimal64 0.2 printed as decimal:    %.25Df
_Decimal64 0.2 printed as scientific: %.25DE
```

In both examples above, libdfp is linked dynamically. When linking a static libdfp to your example, all of libdfp will be discarded by the linker unless you instruct the linker to keep the whole libdfp archive. In this case, something needs to call the `extern int register_printf_dfp (void);` symbol, and you will also need to satisfy the libm dependency. I recommend using the shared library, especially on amd64.

**Comment by newbie-02** [comment]
hello @pmur, thanks for your help,  
  
working better on your machine points to 'there must be some shortcoming  
on mine', will try to find.  
OTOH I have two different machines ( sticks, same hardware ), one Kali,  
one ubuntu, with different ver. of libdfp, both with same problem ...  
  
It's a challenging task to provide info about a 'historically grown' machine,  
System I: Kali Linux ( debian based )  
```  
└─$ cat /etc/os-release 
PRETTY_NAME="Kali GNU/Linux Rolling"
NAME="Kali GNU/Linux"
VERSION_ID="2023.2"
VERSION="2023.2"
VERSION_CODENAME=kali-rolling
```
the kernel is older than the system reg. some problems between newer  
kernels and synaptics input devices on Lenovo notebooks,  
```  
└─$ uname -a  
Linux kali 5.14.0-kali4-amd64 #1 SMP Debian 5.14.16-1kali1 (2021-11-05) x86_64 GNU/Linux  
```  
I have a libdfp directory with a downloaded source, compiled and installed  
some time ago, where the NEWS file announces 1.0.10,  
think to recall that 'sudo apt-get install libdfp-devel' or ~dev returned 'unable to locate',  
```
└─$ ldconfig -p | grep libdfp
        libdfp.so.1 (libc6,x86-64) => /usr/local/lib/libdfp.so.1
        libdfp.so (libc6,x86-64) => /usr/local/lib/libdfp.so
```
gcc:  
```
└─$ gcc --version  
(Debian 11.3.0-12) 11.3.0
```  

System II: ubuntu ( debian based )  
```  
└─$ cat /etc/os-release  
PRETTY_NAME="Ubuntu 22.04.2 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.2 LTS (Jammy Jellyfish)"  
```
here a fresh kernel works,  
```  
└─$ uname -a                    
Linux ubuntu 5.19.0-32-generic #33~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Mon Jan 30 17:03:34 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
```  
I have a libdfp directory with a downloaded source, compiled and installed  
some time ago, where the NEWS file announces 1.0.10,  
think to recall that 'sudo apt-get install libdfp-devel' or ~dev returned 'unable to locate',  
and additional a libdfp-dev: 
```
└─$ sudo apt-get install libdfp-dev 
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
libdfp-dev is already the newest version (1.0.16-1).  
```
gcc:  
```
└─$ gcc --version            
gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
```  
the failing build was invoked on both systems as described in the test program by:  
`gcc -o test1 test1.c -ldfp` and run by ./test1,  
the working one just the same only activated lines marked III. and IV.  
  
do you have any hint where it's best to seek?

**Comment by newbie-02** [comment]
just saw that my libdfp dir showing 1.0.10 in NEWS is "Your branch is up to date with 'origin/debian/master'",  
but https://github.com/libdfp/libdfp/releases announces 1.0.16 as 'actual'.  
Options, need or expect benefit from updating? Or w/could it break system integrity?  
Or is it just an outdated NEWS file?

**Comment by pmur** [comment]
Hi @newbie-02, the NEWS file has not been updated in as long. These days the release notes are contained in the generated ChangeLog file shipped with the release tarballs (which itself is generated from git commit comments).

I'll open a separate issue to clean it up.

**Comment by newbie-02** [comment]
After changing some things to come free from 'lquadmath'  
and have sufficient '-I /usr/local/include/dfp' compilation I still  
experience cases where 'D' printf formatting isn't enabled.  
  
Below a program trying to boil down, and enable check  
of reproducibility. Posted not against problems here, but  
collect basic knowledge also for others.  
```  
// dec64_printing.c, program to demonstrate gcc's _Decimalxxx printing capabilities, 
// compile with 'gcc -o dec64_printing dec64_printing.c', 
// to activate the _Decimalxxx printf formatters 
// compile with 'gcc -o dec64_printing dec64_printing.c -ldfp', 
// and 'un-outcomment' the row starting with '///' to have some decimal calculation active. 
// run with './dec64_printing, 
#define __STDC_WANT_DEC_FP__ 
#include <stdio.h> 									// reg. e.g. printf, 
 
int main() 
{ 
	volatile _Decimal64 x1DD, x2DD; 
	_Decimal64 x3DD; 
	x1DD = 0.2DD; 
	x2DD = 0.1DD; 
 
// print formatting, I had massive problems getting them to work, assume an effect of 'compiler cheating' 
// and not including libdfp.a when substituting all _Decimal calculations by constants at compilation 
// despite needed for printing: 
	printf( "here we try to use print formatting for _Decimal64 values: \n" ); 
	printf( "it requires to link against 'libdfp.a' by adding -ldfp at the compile command. \n" ); 
	printf( "the value %.2Df is represented in decimal64 by                         : %.16Df \n", x1DD, x1DD ); 
	printf( "the value %.2Df is represented in decimal64 by                         : %.16Df \n", x2DD, x2DD ); 
// the following line enforces a decimal calculation at runtime, for which libdfp is included and enables 'D' print formatters, 
/// 	printf( "adding them yields                                                     : %.16Df \n", ( x1DD + x2DD ) ); 
// _assume_ the following calculation is substituted by a constant at compile time, and libdfp is neglected, 
	printf( "for a simple calculation with constants the formatting doesn't work    : %.19Df \n", ( 0.2DD + 0.1DD ) ); 
	printf( "but the calculation is performed with decimal precision                : %.19f \n", (double)( 0.2DD + 0.1DD ) ); 
	printf( "But may be a special of my system, pmur say's it's working well, \n" ); 
	printf( "and no active calculation is required to make 'D' print format working. \n" ); 
	printf( " \n" ); 
} 
```

**Comment by pinskia** [comment]
Oh ubuntu's gcc defaults to --as-needed so if no functions are used from the shared library, it won't be linked.

Try -Wl,--no-as-needed,-ldfp,--as-needed . That will force the linking of the shared library.
